File: scripts/export-results.ts
#!/usr/bin/env node
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { Command } from "commander";
import { stringify } from "csv-stringify/sync";
import fg from "fast-glob";

type JsonObject = Record<string, unknown>;
type Numeric = number | "";

export type ExportOptions = {
  jobDir: string;
  runId?: string;
  outDir?: string;
};

export type TrialRow = {
  run_id: string;
  job_name: string;
  trial_name: string;
  task_name: string;
  task_family: string;
  profile: string;
  agent: string;
  model: string;
  attempt_index: number;
  reward: Numeric;
  passed: boolean;
  exception_type: string;
  exception_message: string;
  started_at: string;
  finished_at: string;
  duration_sec: Numeric;
  environment_setup_duration_sec: Numeric;
  agent_setup_duration_sec: Numeric;
  agent_execution_duration_sec: Numeric;
  verifier_duration_sec: Numeric;
  input_tokens: Numeric;
  cache_tokens: Numeric;
  output_tokens: Numeric;
  cost_usd: Numeric;
  task_checksum: string;
  git_sha: string;
  docs_lock_sha: string;
  result_path: string;
  rewards: Record<string, number>;
};

type SummaryRow = {
  run_id: string;
  job_name: string;
  model: string;
  agent: string;
  task_name: string;
  task_family: string;
  profile: string;
  n_trials: number;
  n_passed: number;
  pass_rate: number;
  n_errors: number;
  mean_reward: Numeric;
  input_tokens: number;
  cache_tokens: number;
  output_tokens: number;
  cost_usd: number;
};

type Metadata = {
  run_id?: string;
  git_sha?: string | null;
  docs_lock?: { sha?: string };
};

type ExportContext = {
  runId: string;
  jobName: string;
  gitSha: string;
  docsLockSha: string;
};

const PROFILES = ["base", "docs", "mcp"] as const;

const TRIAL_COLUMNS = [
  "run_id",
  "job_name",
  "trial_name",
  "task_name",
  "task_family",
  "profile",
  "agent",
  "model",
  "attempt_index",
  "reward",
  "passed",
  "exception_type",
  "exception_message",
  "started_at",
  "finished_at",
  "duration_sec",
  "environment_setup_duration_sec",
  "agent_setup_duration_sec",
  "agent_execution_duration_sec",
  "verifier_duration_sec",
  "input_tokens",
  "cache_tokens",
  "output_tokens",
  "cost_usd",
  "task_checksum",
  "git_sha",
  "docs_lock_sha",
  "result_path",
] as const;

const SUMMARY_COLUMNS = [
  "run_id",
  "job_name",
  "model",
  "agent",
  "task_name",
  "task_family",
  "profile",
  "n_trials",
  "n_passed",
  "pass_rate",
  "n_errors",
  "mean_reward",
  "input_tokens",
  "cache_tokens",
  "output_tokens",
  "cost_usd",
] as const;

function readJson<T extends JsonObject = JsonObject>(filePath: string): T | undefined {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8")) as T;
  } catch {
    return undefined;
  }
}

function readMetadata(jobDir: string): Metadata {
  return readJson<Metadata>(path.join(path.dirname(jobDir), "metadata.json")) ?? {};
}

function commandOutput(command: string, args: string[]): string {
  const result = spawnSync(command, args, { encoding: "utf8" });
  return result.status === 0 ? result.stdout.trim() : "";
}

function docsLockSha(): string {
  return readJson(path.join("config", "tempo-docs.lock.json"))?.sha?.toString() ?? "";
}

function defaultRunId(jobDir: string, metadata: Metadata): string {
  if (metadata.run_id) return metadata.run_id;
  return path.basename(jobDir) === "harbor-job"
    ? path.basename(path.dirname(jobDir))
    : path.basename(jobDir);
}

function defaultOutDir(jobDir: string): string {
  return path.basename(jobDir) === "harbor-job"
    ? path.join(path.dirname(jobDir), "exports")
    : path.join(jobDir, "exports");
}

function buildContext(jobDir: string, runId: string, metadata: Metadata): ExportContext {
  return {
    runId,
    jobName: path.basename(jobDir),
    gitSha: metadata.git_sha ?? commandOutput("git", ["rev-parse", "HEAD"]),
    docsLockSha: metadata.docs_lock?.sha ?? docsLockSha(),
  };
}

function resultFiles(root: string): string[] {
  return fg
    .sync("**/result.json", {
      absolute: true,
      cwd: root,
      ignore: ["**/exports/**"],
      onlyFiles: true,
    })
    .sort();
}

export function parseTaskName(taskName: string): { taskFamily: string; profile: string } {
  for (const profile of PROFILES) {
    const suffix = `-${profile}`;
    if (taskName.endsWith(suffix)) {
      return { taskFamily: taskName.slice(0, -suffix.length), profile };
    }
  }
  return { taskFamily: taskName, profile: "unknown" };
}

function asObject(value: unknown): JsonObject {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as JsonObject)
    : {};
}

function asNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function asString(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function durationSec(timing: unknown): Numeric {
  const object = asObject(timing);
  const startedAt = asString(object.started_at);
  const finishedAt = asString(object.finished_at);
  if (!startedAt || !finishedAt) return "";

  const started = Date.parse(startedAt);
  const finished = Date.parse(finishedAt);
  if (!Number.isFinite(started) || !Number.isFinite(finished)) return "";
  return Math.max((finished - started) / 1000, 0);
}

function tokenTotals(result: JsonObject): {
  input: Numeric;
  cache: Numeric;
  output: Numeric;
  cost: Numeric;
} {
  const contexts = collectAgentContexts(result);
  return {
    input: sumOptional(contexts, "n_input_tokens"),
    cache: sumOptional(contexts, "n_cache_tokens"),
    output: sumOptional(contexts, "n_output_tokens"),
    cost: sumOptional(contexts, "cost_usd"),
  };
}

function collectAgentContexts(result: JsonObject): JsonObject[] {
  const contexts: JsonObject[] = [];
  const agentResult = asObject(result.agent_result);
  if (Object.keys(agentResult).length > 0) contexts.push(agentResult);

  for (const step of Array.isArray(result.step_results) ? result.step_results : []) {
    const stepAgentResult = asObject(asObject(step).agent_result);
    if (Object.keys(stepAgentResult).length > 0) contexts.push(stepAgentResult);
  }
  return contexts;
}

function sumOptional(objects: JsonObject[], key: string): Numeric {
  let total = 0;
  let found = false;
  for (const object of objects) {
    const value = asNumber(object[key]);
    if (value === undefined) continue;
    total += value;
    found = true;
  }
  return found ? total : "";
}

function extractRewards(result: JsonObject): Record<string, number> {
  const rawRewards = asObject(asObject(result.verifier_result).rewards);
  return Object.fromEntries(
    Object.entries(rawRewards).flatMap(([key, value]) => {
      const numberValue = asNumber(value);
      return numberValue === undefined ? [] : [[key, numberValue]];
    }),
  );
}

function primaryReward(rewards: Record<string, number>): Numeric {
  if (typeof rewards.reward === "number") return rewards.reward;
  if (typeof rewards.score === "number") return rewards.score;
  const values = Object.values(rewards);
  return values.length === 1 ? values[0] : "";
}

function parseTrialResult(filePath: string, context: ExportContext): TrialRow | undefined {
  const result = readJson(filePath);
  if (!result || typeof result.task_name !== "string" || typeof result.trial_name !== "string") {
    return undefined;
  }

  const agentInfo = asObject(result.agent_info);
  const modelInfo = asObject(agentInfo.model_info);
  const exceptionInfo = asObject(result.exception_info);
  const task = parseTaskName(result.task_name);
  const rewards = extractRewards(result);
  const reward = primaryReward(rewards);
  const tokens = tokenTotals(result);

  return {
    run_id: context.runId,
    job_name: context.jobName,
    trial_name: result.trial_name,
    task_name: result.task_name,
    task_family: task.taskFamily,
    profile: task.profile,
    agent: asString(agentInfo.name),
    model: asString(modelInfo.name),
    attempt_index: 0,
    reward,
    passed: typeof reward === "number" && reward >= 1,
    exception_type: asString(exceptionInfo.exception_type),
    exception_message: asString(exceptionInfo.exception_message),
    started_at: asString(result.started_at),
    finished_at: asString(result.finished_at),
    duration_sec: durationSec({
      started_at: result.started_at,
      finished_at: result.finished_at,
    }),
    environment_setup_duration_sec: durationSec(result.environment_setup),
    agent_setup_duration_sec: durationSec(result.agent_setup),
    agent_execution_duration_sec: durationSec(result.agent_execution),
    verifier_duration_sec: durationSec(result.verifier),
    input_tokens: tokens.input,
    cache_tokens: tokens.cache,
    output_tokens: tokens.output,
    cost_usd: tokens.cost,
    task_checksum: asString(result.task_checksum),
    git_sha: context.gitSha,
    docs_lock_sha: context.docsLockSha,
    result_path: filePath,
    rewards,
  };
}

function assignAttemptIndexes(rows: TrialRow[]): TrialRow[] {
  const counts = new Map<string, number>();
  return rows
    .sort((a, b) => a.trial_name.localeCompare(b.trial_name))
    .map((row) => {
      const key = [row.task_name, row.agent, row.model].join("\0");
      const attemptIndex = (counts.get(key) ?? 0) + 1;
      counts.set(key, attemptIndex);
      return { ...row, attempt_index: attemptIndex };
    });
}

function rewardColumns(rows: TrialRow[]): string[] {
  return [...new Set(rows.flatMap((row) => Object.keys(row.rewards)))]
    .sort()
    .map((key) => `verifier_reward_${key}`);
}

function trialCsvRows(rows: TrialRow[]): Record<string, unknown>[] {
  return rows.map(({ rewards, ...row }) => ({
    ...row,
    ...Object.fromEntries(
      Object.entries(rewards).map(([key, value]) => [`verifier_reward_${key}`, value]),
    ),
  }));
}

function writeCsv(filePath: string, columns: string[], rows: Record<string, unknown>[]) {
  fs.writeFileSync(filePath, stringify(rows, { columns, header: true }));
}

function sumNumbers(values: Numeric[]): number {
  let sum = 0;
  for (const value of values) {
    if (typeof value === "number") sum += value;
  }
  return sum;
}

function groupRows(rows: TrialRow[]): Map<string, TrialRow[]> {
  const groups = new Map<string, TrialRow[]>();
  for (const row of rows) {
    const key = [row.model, row.agent, row.task_name, row.task_family, row.profile].join("\0");
    groups.set(key, [...(groups.get(key) ?? []), row]);
  }
  return groups;
}

export function summarizeRows(rows: TrialRow[]): SummaryRow[] {
  return [...groupRows(rows).values()]
    .map((group) => summarizeGroup(group))
    .sort((a, b) =>
      [a.model, a.task_name, a.profile].join("\0").localeCompare(
        [b.model, b.task_name, b.profile].join("\0"),
      ),
    );
}

function summarizeGroup(group: TrialRow[]): SummaryRow {
  const first = group[0];
  const rewards = group
    .map((row) => row.reward)
    .filter((value): value is number => typeof value === "number");
  const nPassed = group.filter((row) => row.passed).length;

  return {
    run_id: first.run_id,
    job_name: first.job_name,
    model: first.model,
    agent: first.agent,
    task_name: first.task_name,
    task_family: first.task_family,
    profile: first.profile,
    n_trials: group.length,
    n_passed: nPassed,
    pass_rate: group.length === 0 ? 0 : nPassed / group.length,
    n_errors: group.filter((row) => row.exception_type).length,
    mean_reward:
      rewards.length === 0
        ? ""
        : rewards.reduce((sum, value) => sum + value, 0) / rewards.length,
    input_tokens: sumNumbers(group.map((row) => row.input_tokens)),
    cache_tokens: sumNumbers(group.map((row) => row.cache_tokens)),
    output_tokens: sumNumbers(group.map((row) => row.output_tokens)),
    cost_usd: sumNumbers(group.map((row) => row.cost_usd)),
  };
}

export function exportResults(options: ExportOptions): {
  runId: string;
  outDir: string;
  trials: TrialRow[];
  summary: SummaryRow[];
} {
  const jobDir = path.resolve(options.jobDir);
  if (!fs.existsSync(jobDir)) {
    throw new Error(`Job directory not found: ${jobDir}`);
  }

  const metadata = readMetadata(jobDir);
  const runId = options.runId ?? defaultRunId(jobDir, metadata);
  const outDir = path.resolve(options.outDir ?? defaultOutDir(jobDir));
  const context = buildContext(jobDir, runId, metadata);
  const trials = assignAttemptIndexes(
    resultFiles(jobDir)
      .map((filePath) => parseTrialResult(filePath, context))
      .filter((row): row is TrialRow => row !== undefined),
  );
  const summary = summarizeRows(trials);

  fs.mkdirSync(outDir, { recursive: true });
  writeCsv(
    path.join(outDir, "trials.csv"),
    [...TRIAL_COLUMNS, ...rewardColumns(trials)],
    trialCsvRows(trials),
  );
  writeCsv(path.join(outDir, "summary.csv"), [...SUMMARY_COLUMNS], summary);
  writeSummaryJson(path.join(outDir, "summary.json"), jobDir, runId, trials, summary);

  return { runId, outDir, trials, summary };
}

function writeSummaryJson(
  filePath: string,
  jobDir: string,
  runId: string,
  trials: TrialRow[],
  summary: SummaryRow[],
) {
  fs.writeFileSync(
    filePath,
    `${JSON.stringify(
      {
        schema_version: 1,
        run_id: runId,
        job_dir: jobDir,
        exported_at: new Date().toISOString(),
        n_trials: trials.length,
        reward_keys: [...new Set(trials.flatMap((row) => Object.keys(row.rewards)))].sort(),
        summary,
      },
      null,
      2,
    )}\n`,
  );
}

function cliOptions(argv: string[]): ExportOptions {
  const program = new Command()
    .name("results:export")
    .description("Export Harbor trial results to CSV and JSON.")
    .requiredOption("--job <path>", "Harbor job directory, for example runs/<run_id>/harbor-job")
    .option("--run-id <name>", "Run id to write into exports")
    .option("--out-dir <path>", "Output directory, defaults to runs/<run_id>/exports")
    .showHelpAfterError();

  program.parse(argv);
  const options = program.opts<{ job: string; runId?: string; outDir?: string }>();
  return {
    jobDir: options.job,
    runId: options.runId,
    outDir: options.outDir,
  };
}

const currentFile = fileURLToPath(import.meta.url);
if (process.argv[1] && path.resolve(process.argv[1]) === currentFile) {
  try {
    const result = exportResults(cliOptions(process.argv));
    console.log(`Exported ${result.trials.length} trials to ${result.outDir}`);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

File: scripts/export_results.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import subprocess
from datetime import datetime
from pathlib import Path
from typing import Any

import pandas as pd

JsonObject = dict[str, Any]
Numeric = float | int | str

MCP_PROFILE_SUFFIX = "-mcp"

TRIAL_COLUMNS = [
    "run_id",
    "job_name",
    "trial_name",
    "task_name",
    "task_family",
    "profile",
    "agent",
    "model",
    "attempt_index",
    "reward",
    "passed",
    "exception_type",
    "exception_message",
    "started_at",
    "finished_at",
    "duration_sec",
    "environment_setup_duration_sec",
    "agent_setup_duration_sec",
    "agent_execution_duration_sec",
    "verifier_duration_sec",
    "input_tokens",
    "cache_tokens",
    "output_tokens",
    "cost_usd",
    "task_checksum",
    "git_sha",
    "docs_source",
    "result_path",
]

SUMMARY_COLUMNS = [
    "run_id",
    "job_name",
    "model",
    "agent",
    "task_name",
    "task_family",
    "profile",
    "n_trials",
    "n_passed",
    "pass_rate",
    "n_errors",
    "mean_reward",
    "input_tokens",
    "cache_tokens",
    "output_tokens",
    "cost_usd",
]


def read_json(file_path: Path) -> JsonObject | None:
    try:
        return json.loads(file_path.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return None


def read_metadata(job_dir: Path) -> JsonObject:
    return read_json(job_dir.parent / "metadata.json") or {}


def command_output(command: str, args: list[str]) -> str:
    result = subprocess.run(
        [command, *args],
        check=False,
        capture_output=True,
        text=True,
    )
    return result.stdout.strip() if result.returncode == 0 else ""


def default_docs_source() -> str:
    sha = (read_json(Path("config/tempo-docs.lock.json")) or {}).get("sha")
    return str(sha) if sha else "public"


def default_run_id(job_dir: Path, metadata: JsonObject) -> str:
    if metadata.get("run_id"):
        return str(metadata["run_id"])
    return job_dir.parent.name if job_dir.name == "harbor-job" else job_dir.name


def default_out_dir(job_dir: Path) -> Path:
    return (
        job_dir.parent / "exports"
        if job_dir.name == "harbor-job"
        else job_dir / "exports"
    )


def build_context(job_dir: Path, run_id: str, metadata: JsonObject) -> JsonObject:
    source = metadata.get("docs_source")
    return {
        "run_id": run_id,
        "job_name": job_dir.name,
        "git_sha": metadata.get("git_sha")
        or command_output("git", ["rev-parse", "HEAD"]),
        "docs_source": source
        if isinstance(source, str) and source
        else default_docs_source(),
        "tempo_profile": profile_from_job(job_dir),
    }


def result_files(root: Path) -> list[Path]:
    return sorted(
        path
        for path in root.rglob("result.json")
        if "exports" not in path.relative_to(root).parts
    )


def profile_from_job(job_dir: Path) -> str:
    config = read_json(job_dir / "config.json") or {}
    for agent in as_object(config).get("agents", []):
        if not isinstance(agent, dict):
            continue
        for server in agent.get("mcp_servers") or []:
            if isinstance(server, dict) and server.get("name") == "tempo":
                return "mcp"
    return "docs"


def parse_task_name(task_name: str, tempo_profile: str = "docs") -> dict[str, str]:
    if task_name.endswith(MCP_PROFILE_SUFFIX):
        return {"task_family": task_name[: -len(MCP_PROFILE_SUFFIX)], "profile": "mcp"}
    if task_name.startswith("tempo-v1/"):
        return {"task_family": task_name, "profile": tempo_profile}
    return {"task_family": task_name, "profile": "unknown"}


def as_object(value: Any) -> JsonObject:
    return value if isinstance(value, dict) else {}


def as_number(value: Any) -> float | int | None:
    return (
        value
        if isinstance(value, int | float) and not isinstance(value, bool)
        else None
    )


def as_string(value: Any) -> str:
    return value if isinstance(value, str) else ""


def parse_time(value: str) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None


def duration_sec(timing: Any) -> Numeric:
    timing_obj = as_object(timing)
    started = parse_time(as_string(timing_obj.get("started_at")))
    finished = parse_time(as_string(timing_obj.get("finished_at")))
    if not started or not finished:
        return ""
    return max((finished - started).total_seconds(), 0)


def collect_agent_contexts(result: JsonObject) -> list[JsonObject]:
    contexts: list[JsonObject] = []
    agent_result = as_object(result.get("agent_result"))
    if agent_result:
        contexts.append(agent_result)
    for step in (
        result.get("step_results")
        if isinstance(result.get("step_results"), list)
        else []
    ):
        step_agent_result = as_object(as_object(step).get("agent_result"))
        if step_agent_result:
            contexts.append(step_agent_result)
    return contexts


def sum_optional(objects: list[JsonObject], key: str) -> Numeric:
    total: float | int = 0
    found = False
    for obj in objects:
        value = as_number(obj.get(key))
        if value is None:
            continue
        total += value
        found = True
    return total if found else ""


def token_totals(result: JsonObject) -> dict[str, Numeric]:
    contexts = collect_agent_contexts(result)
    return {
        "input": sum_optional(contexts, "n_input_tokens"),
        "cache": sum_optional(contexts, "n_cache_tokens"),
        "output": sum_optional(contexts, "n_output_tokens"),
        "cost": sum_optional(contexts, "cost_usd"),
    }


def extract_rewards(result: JsonObject) -> dict[str, float | int]:
    raw_rewards = as_object(as_object(result.get("verifier_result")).get("rewards"))
    return {
        key: value for key, value in raw_rewards.items() if as_number(value) is not None
    }


def primary_reward(rewards: dict[str, float | int]) -> Numeric:
    if isinstance(rewards.get("reward"), int | float):
        return rewards["reward"]
    if isinstance(rewards.get("score"), int | float):
        return rewards["score"]
    values = list(rewards.values())
    return values[0] if len(values) == 1 else ""


def parse_trial_result(file_path: Path, context: JsonObject) -> JsonObject | None:
    result = read_json(file_path)
    if not result or not isinstance(result.get("task_name"), str):
        return None
    if not isinstance(result.get("trial_name"), str):
        return None

    agent_info = as_object(result.get("agent_info"))
    model_info = as_object(agent_info.get("model_info"))
    exception_info = as_object(result.get("exception_info"))
    task = parse_task_name(result["task_name"], as_string(context.get("tempo_profile")))
    rewards = extract_rewards(result)
    reward = primary_reward(rewards)
    tokens = token_totals(result)

    return {
        "run_id": context["run_id"],
        "job_name": context["job_name"],
        "trial_name": result["trial_name"],
        "task_name": result["task_name"],
        "task_family": task["task_family"],
        "profile": task["profile"],
        "agent": as_string(agent_info.get("name")),
        "model": as_string(model_info.get("name")),
        "attempt_index": 0,
        "reward": reward,
        "passed": isinstance(reward, int | float) and reward >= 1,
        "exception_type": as_string(exception_info.get("exception_type")),
        "exception_message": as_string(exception_info.get("exception_message")),
        "started_at": as_string(result.get("started_at")),
        "finished_at": as_string(result.get("finished_at")),
        "duration_sec": duration_sec(
            {
                "started_at": result.get("started_at"),
                "finished_at": result.get("finished_at"),
            },
        ),
        "environment_setup_duration_sec": duration_sec(result.get("environment_setup")),
        "agent_setup_duration_sec": duration_sec(result.get("agent_setup")),
        "agent_execution_duration_sec": duration_sec(result.get("agent_execution")),
        "verifier_duration_sec": duration_sec(result.get("verifier")),
        "input_tokens": tokens["input"],
        "cache_tokens": tokens["cache"],
        "output_tokens": tokens["output"],
        "cost_usd": tokens["cost"],
        "task_checksum": as_string(result.get("task_checksum")),
        "git_sha": context["git_sha"],
        "docs_source": context["docs_source"],
        "result_path": str(file_path),
        "rewards": rewards,
    }


def assign_attempt_indexes(rows: list[JsonObject]) -> list[JsonObject]:
    counts: dict[str, int] = {}
    sorted_rows = sorted(rows, key=lambda row: str(row["trial_name"]))
    for row in sorted_rows:
        key = "\0".join([str(row["task_name"]), str(row["agent"]), str(row["model"])])
        counts[key] = counts.get(key, 0) + 1
        row["attempt_index"] = counts[key]
    return sorted_rows


def reward_columns(rows: list[JsonObject]) -> list[str]:
    keys = sorted({key for row in rows for key in row["rewards"]})
    return [f"verifier_reward_{key}" for key in keys]


def trial_csv_rows(rows: list[JsonObject]) -> list[JsonObject]:
    csv_rows = []
    for row in rows:
        csv_row = {key: value for key, value in row.items() if key != "rewards"}
        csv_row.update(
            {f"verifier_reward_{key}": value for key, value in row["rewards"].items()},
        )
        csv_rows.append(csv_row)
    return csv_rows


def csv_value(value: Any) -> Any:
    if isinstance(value, bool):
        return str(value).lower()
    if value is None:
        return ""
    return value


def write_csv(file_path: Path, columns: list[str], rows: list[JsonObject]) -> None:
    frame = pd.DataFrame(
        [{key: csv_value(row.get(key, "")) for key in columns} for row in rows],
        columns=columns,
    )
    frame.to_csv(file_path, index=False)


def summarize_rows(rows: list[JsonObject]) -> list[JsonObject]:
    if not rows:
        return []

    frame = pd.DataFrame(rows)
    frame["_reward"] = pd.to_numeric(frame["reward"], errors="coerce")
    frame["_passed"] = frame["passed"].astype(int)
    frame["_error"] = frame["exception_type"].astype(bool).astype(int)
    for column in ["input_tokens", "cache_tokens", "output_tokens", "cost_usd"]:
        frame[column] = pd.to_numeric(frame[column], errors="coerce").fillna(0)

    group_columns = ["model", "agent", "task_name", "task_family", "profile"]
    summary = (
        frame.groupby(group_columns, as_index=False, sort=False)
        .agg(
            run_id=("run_id", "first"),
            job_name=("job_name", "first"),
            n_trials=("trial_name", "size"),
            n_passed=("_passed", "sum"),
            n_errors=("_error", "sum"),
            mean_reward=("_reward", "mean"),
            input_tokens=("input_tokens", "sum"),
            cache_tokens=("cache_tokens", "sum"),
            output_tokens=("output_tokens", "sum"),
            cost_usd=("cost_usd", "sum"),
        )
        .assign(pass_rate=lambda data: data["n_passed"] / data["n_trials"])
        .sort_values(["model", "task_name", "profile"], kind="stable")
    )
    summary["mean_reward"] = summary["mean_reward"].where(
        summary["mean_reward"].notna(),
        "",
    )
    return summary[SUMMARY_COLUMNS].to_dict("records")


def write_summary_json(
    file_path: Path,
    job_dir: Path,
    run_id: str,
    trials: list[JsonObject],
    summary: list[JsonObject],
) -> None:
    reward_keys = sorted({key for row in trials for key in row["rewards"]})
    file_path.write_text(
        f"{
            json.dumps(
                {
                    'schema_version': 1,
                    'run_id': run_id,
                    'job_dir': str(job_dir),
                    'exported_at': datetime.now().astimezone().isoformat(),
                    'n_trials': len(trials),
                    'reward_keys': reward_keys,
                    'summary': summary,
                },
                indent=2,
            )
        }\n",
    )


def export_results(
    job_dir: str | Path,
    run_id: str | None = None,
    out_dir: str | Path | None = None,
) -> dict[str, Any]:
    job_path = Path(job_dir).resolve()
    if not job_path.exists():
        msg = f"Job directory not found: {job_path}"
        raise FileNotFoundError(msg)

    metadata = read_metadata(job_path)
    effective_run_id = run_id or default_run_id(job_path, metadata)
    out_path = (
        Path(out_dir).resolve() if out_dir else default_out_dir(job_path).resolve()
    )
    context = build_context(job_path, effective_run_id, metadata)
    trials = assign_attempt_indexes(
        [
            row
            for file_path in result_files(job_path)
            if (row := parse_trial_result(file_path, context)) is not None
        ],
    )
    summary = summarize_rows(trials)

    out_path.mkdir(parents=True, exist_ok=True)
    write_csv(
        out_path / "trials.csv",
        [*TRIAL_COLUMNS, *reward_columns(trials)],
        trial_csv_rows(trials),
    )
    write_csv(out_path / "summary.csv", SUMMARY_COLUMNS, summary)
    write_summary_json(
        out_path / "summary.json", job_path, effective_run_id, trials, summary
    )

    return {
        "run_id": effective_run_id,
        "out_dir": str(out_path),
        "trials": trials,
        "summary": summary,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="results:export",
        description="Export Harbor trial results to CSV and JSON.",
    )
    parser.add_argument(
        "--job",
        required=True,
        help="Harbor job directory, for example runs/<run_id>/harbor-job",
    )
    parser.add_argument("--run-id", help="Run id to write into exports")
    parser.add_argument("--out-dir", help="Output directory")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    result = export_results(args.job, run_id=args.run_id, out_dir=args.out_dir)
    print(f"Exported {len(result['trials'])} trials to {result['out_dir']}")


if __name__ == "__main__":
    main()
